{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "577cec55",
   "metadata": {},
   "source": [
    "# Tutorial 2: ATAC-RNA-Seq Mouse Brain RNA and ATAC Multi-Omics Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d0b44ca",
   "metadata": {},
   "source": [
    "We applied COSMOS to analyze a spatially resolved multi-omics dataset based on real-world experiments instead of simulated data. This experimental dataset contains P22 mouse brain coronal sections with a joint profiling of spatial chromatin accessibility (ATAC) and spatial transcriptome (RNA). The annotation was manually generated by aligning the sample with the P56 mouse brain coronal section from Allen Mouse Brain Atlas.\n",
    "\n",
    "The raw data can be downloaded from: \n",
    "https://brain-spatial-omics.cells.ucsc.edu\n",
    "\n",
    "The processed data is available at: https://zenodo.org/records/13932144"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef369d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from umap import UMAP\n",
    "import sklearn\n",
    "import sklearn.metrics\n",
    "import seaborn as sns\n",
    "# from COSMOS.cosmos import Cosmos\n",
    "from COSMOS.cosmos_transformer_pe_version import Cosmos\n",
    "from COSMOS.pyWNN import pyWNN \n",
    "import h5py\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "random_seed = 20"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "478b1e45",
   "metadata": {},
   "source": [
    "##  Preparation of data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b413659f",
   "metadata": {},
   "source": [
    "### Importing the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9049b5ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_mat = h5py.File('./ATAC_RNA_Seq_MouseBrain_RNA_ATAC.h5', 'r')\n",
    "df_data_RNA = np.array(data_mat['X_RNA']).astype('float64')     # gene count matrix\n",
    "df_data_ATAC= np.array(data_mat['X_ATAC']).astype('float64')  # protein count matrix\n",
    "loc = np.array(data_mat['Pos']).astype('float64') \n",
    "LayerName = list(data_mat['LayerName'])\n",
    "LayerName = [item.decode(\"utf-8\") for item in LayerName]\n",
    "\n",
    "adata1 = sc.AnnData(df_data_RNA, dtype=\"float64\")\n",
    "adata1.obsm['spatial'] = np.array(loc)\n",
    "adata1.obs['LayerName'] = LayerName\n",
    "adata1.obs['x_pos'] = np.array(loc)[:,0]\n",
    "adata1.obs['y_pos'] = np.array(loc)[:,1]\n",
    "\n",
    "adata2 = sc.AnnData(df_data_ATAC, dtype=\"float64\")\n",
    "adata2.obsm['spatial'] = np.array(loc)\n",
    "adata2.obs['LayerName'] = LayerName\n",
    "adata2.obs['x_pos'] = np.array(loc)[:,0]\n",
    "adata2.obs['y_pos'] = np.array(loc)[:,1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a35e163c",
   "metadata": {},
   "source": [
    "### Visualizing spatial positions with annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79954d85",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_new = adata1.copy()\n",
    "adata_new.obs[\"LayerName\"]=adata_new.obs[\"LayerName\"].astype('category')\n",
    "\n",
    "\n",
    "matplotlib.rcParams['font.size'] = 12.0\n",
    "fig, axes = plt.subplots(1, 1, figsize=(3.5,3))\n",
    "sz = 20\n",
    "plot_color=['#D1D1D1','#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', \\\n",
    "            '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#ffd8b1', '#800000', '#aaffc3', '#808000', '#000075', '#000000', '#808080', '#ffffff', '#fffac8']\n",
    "\n",
    "domains=\"LayerName\"\n",
    "num_celltype=len(adata_new.obs[domains].unique())\n",
    "adata_new.uns[domains+\"_colors\"]=list(plot_color[:num_celltype])\n",
    "titles = 'Annotation ' \n",
    "ax=sc.pl.scatter(adata_new,alpha=1,x=\"x_pos\",y=\"y_pos\",color=domains,title=titles ,color_map=plot_color,show=False,size=sz,ax = axes)\n",
    "ax.axis('off')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "592c8dcf",
   "metadata": {},
   "source": [
    "## Applying COSMOS to integrate RNA and ATAC omics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5482063d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## COSMOS training \n",
    "cosmos_comb = Cosmos(adata1=adata1,adata2=adata2)\n",
    "cosmos_comb.preprocessing_data(n_neighbors = 10)\n",
    "cosmos_comb.train(spatial_regularization_strength=0.01, z_dim=50, \n",
    "         lr=1e-3, wnn_epoch = 100, total_epoch=1000, max_patience_bef=10, max_patience_aft=30, min_stop=200, \n",
    "         random_seed=random_seed, gpu=0, regularization_acceleration=True, edge_subset_sz=1000000)\n",
    "weights = cosmos_comb.weights\n",
    "df_embedding = pd.DataFrame(cosmos_comb.embedding)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94c47bcb",
   "metadata": {},
   "source": [
    "### Clustering of COSMOS integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d1ee183",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_new = adata1.copy()\n",
    "embedding_adata = sc.AnnData(df_embedding)\n",
    "sc.pp.neighbors(embedding_adata, n_neighbors=50, use_rep='X')\n",
    "\n",
    "# Manualy setting resolution for clustering\n",
    "res = 1.0\n",
    "sc.tl.leiden(embedding_adata, resolution=res)\n",
    "adata_new.obs['Cluster_cosmos'] = list(embedding_adata.obs[\"leiden\"].cat.codes)\n",
    "adata_new.obs[\"Cluster_cosmos\"]=adata_new.obs[\"Cluster_cosmos\"].astype('category')\n",
    "\n",
    "# Relabeling clusters with layer names\n",
    "digit_labels = list(adata_new.obs['Cluster_cosmos'])\n",
    "for i in range(len(digit_labels)):\n",
    "    if digit_labels[i] not in [0,1,2,4,5,6,7,8,11]:\n",
    "        digit_labels[i] = 100\n",
    "adata_new.obs['Cluster_cosmos'] = digit_labels\n",
    "adata_new.obs[\"Cluster_cosmos\"]=adata_new.obs[\"Cluster_cosmos\"].astype('category')\n",
    "adata_new.obs['Cluster_cosmos']=adata_new.obs['Cluster_cosmos'].cat.rename_categories({0: 'CP2',\n",
    "                                                   1: 'L5',\n",
    "                                                   2: 'CP1',\n",
    "                                                   4: 'L4',\n",
    "                                                   5: 'ccg/aco',\n",
    "                                                   6: 'ACB',\n",
    "                                                   7: 'L1-L3',\n",
    "                                                    8: 'L6a/b',\n",
    "                                                   11: 'VL',\n",
    "                                                    100: '1_others'\n",
    "                                                         })\n",
    "\n",
    "ordered_cluster = np.unique(list(adata_new.obs['Cluster_cosmos']))\n",
    "adata_new.obs['Cluster_cosmos'] = adata_new.obs['Cluster_cosmos'].cat.reorder_categories(ordered_cluster, ordered=True)\n",
    "\n",
    "# Calculating ARI\n",
    "opt_cluster_cosmos = list(adata_new.obs['Cluster_cosmos'])\n",
    "opt_ari_cosmos = sklearn.metrics.adjusted_rand_score(LayerName, opt_cluster_cosmos)\n",
    "opt_ari_cosmos = round(opt_ari_cosmos, 2)\n",
    "\n",
    "# Ploting figures\n",
    "matplotlib.rcParams['font.size'] = 8.0\n",
    "fig, axes = plt.subplots(1, 2, figsize=(6,2))\n",
    "sz = 10\n",
    "plot_color=['#D1D1D1','#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', \\\n",
    "            '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#ffd8b1', '#800000', '#aaffc3', '#808000', '#000075', '#000000', '#808080', '#ffffff', '#fffac8']\n",
    "domains=\"LayerName\"\n",
    "num_celltype=len(adata_new.obs[domains].unique())\n",
    "adata_new.uns[domains+\"_colors\"]=list(plot_color[:num_celltype])\n",
    "titles = 'Annotation' \n",
    "ax=sc.pl.scatter(adata_new,alpha=1,x=\"x_pos\",y=\"y_pos\",color=domains,title=titles ,color_map=plot_color,show=False,size=sz,ax = axes[0])\n",
    "ax.axis('off')\n",
    "\n",
    "plot_color_1=['#D1D1D1','#e6194b', '#3cb44b', '#bcf60c','#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', \\\n",
    "             '#fabebe', '#008080', '#e6beff', '#9a6324', '#ffd8b1', '#800000', '#aaffc3', '#808000', '#000075', '#000000', '#808080', '#ffffff', '#fffac8']\n",
    "domains=\"Cluster_cosmos\"\n",
    "num_celltype=len(adata_new.obs[domains].unique())\n",
    "adata_new.uns[domains+\"_colors\"]=list(plot_color_1[:num_celltype])\n",
    "titles = 'COSMOS, ARI = ' + str(opt_ari_cosmos)\n",
    "ax=sc.pl.scatter(adata_new,alpha=1,x=\"x_pos\",y=\"y_pos\",color=domains,title=titles ,color_map=plot_color_1,show=False,size=sz,ax = axes[1])\n",
    "ax.axis('off')\n",
    "plt.tight_layout()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9137d57",
   "metadata": {},
   "source": [
    "### UMAP visualization of COSMOS integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57cffd41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP visualization\n",
    "umap_2d = UMAP(n_components=2, init='random', random_state=random_seed, min_dist = 0.3,n_neighbors=30)\n",
    "umap_pos = umap_2d.fit_transform(df_embedding)\n",
    "adata_new.obs['cosmos_umap_pos_x'] = umap_pos[:,0]\n",
    "adata_new.obs['cosmos_umap_pos_y'] = umap_pos[:,1]\n",
    "\n",
    "# Ploting figures\n",
    "matplotlib.rcParams['font.size'] = 8.0\n",
    "fig, axes = plt.subplots(1, 1, figsize=(4,3))\n",
    "sz = 10\n",
    "domains=\"Cluster_cosmos\"\n",
    "num_celltype=len(adata_new.obs[domains].unique())\n",
    "adata_new.uns[domains+\"_colors\"]=list(plot_color_1[:num_celltype])\n",
    "titles = 'UMAP of COSMOS'\n",
    "ax=sc.pl.scatter(adata_new,alpha=1,x=\"cosmos_umap_pos_x\",y=\"cosmos_umap_pos_y\",color=domains,title=titles ,color_map=plot_color_1,show=False,size=sz,ax = axes)\n",
    "ax.axis('off')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93b5829c",
   "metadata": {},
   "source": [
    "### Pseudo-spatiotemporal map (pSM) from COSMOS integration\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f96dc6d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Calculating pseudo-times \n",
    "embedding_adata.uns['iroot'] = np.flatnonzero(adata_new.obs[\"Cluster_cosmos\"] == 'CP2')[0]\n",
    "sc.tl.diffmap(embedding_adata)\n",
    "sc.tl.dpt(embedding_adata)\n",
    "pSM_values_cosmos = embedding_adata.obs['dpt_pseudotime'].to_numpy()\n",
    "\n",
    "\n",
    "# Ploting figures\n",
    "matplotlib.rcParams['font.size'] = 8.0\n",
    "fig, axes = plt.subplots(1, 1, figsize=(3,3))\n",
    "sz = 10\n",
    "x = np.array(adata_new.obs['x_pos'])\n",
    "y = np.array(adata_new.obs['y_pos'])\n",
    "ax_temp = axes\n",
    "im = ax_temp.scatter(x, y, s=20, c=pSM_values_cosmos, marker='.', cmap='coolwarm',alpha = 1)\n",
    "ax_temp.axis('off')\n",
    "ax_temp.set_title('dpt_pseudotime of COSMOS')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04d2d303",
   "metadata": {},
   "source": [
    "### Showing modality weights of two omics in COSMOS integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d88795d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_weight_value(alpha, label, modality1='RNA', modality2='ATAC',order = None):\n",
    "    df = pd.DataFrame(columns=[modality1, modality2, 'label'])  \n",
    "    df[modality1], df[modality2] = alpha[:, 0], alpha[:, 1]\n",
    "    df['label'] = label\n",
    "    df = df.set_index('label').stack().reset_index()\n",
    "    df.columns = ['label_COSMOS', 'Modality', 'Modality weights']\n",
    "    matplotlib.rcParams['font.size'] = 8.0\n",
    "    fig, axes = plt.subplots(1, 1, figsize=(5,3))\n",
    "    ax = sns.violinplot(data=df, x='label_COSMOS', y='Modality weights', hue=\"Modality\",\n",
    "                split=True, inner=\"quart\", linewidth=1, orient = 'v', order=order)\n",
    "    plt.tight_layout(w_pad=0.05)\n",
    "    ax.set_title(modality1 + ' vs ' + modality2)\n",
    "    ax.set_xticklabels(order, rotation = 45)\n",
    "ordered_cluster = np.unique(LayerName)\n",
    "layer_type = ordered_cluster\n",
    "index_all = [np.array([i for i in range(len(LayerName)) if LayerName[i] == ordered_cluster[0]])]\n",
    "for k in range(1,len(layer_type)):\n",
    "    temp_idx = np.array([i for i in range(len(LayerName)) if LayerName[i] == ordered_cluster[k]])\n",
    "    index_all.append(temp_idx)\n",
    "    \n",
    "wghts_mean = np.mean(weights[index_all[0],:],0)\n",
    "for k in range(1,len(ordered_cluster)):\n",
    "    wghts_mean_temp = np.mean(weights[index_all[k],:],0)\n",
    "    wghts_mean = np.vstack([wghts_mean, wghts_mean_temp])\n",
    "df_wghts_mean = pd.DataFrame(wghts_mean,columns = ['w1','w2'],index = ordered_cluster)\n",
    "\n",
    "df_sort_mean = df_wghts_mean.sort_values(by=['w1'])\n",
    "plot_weight_value(weights, np.array(adata_new.obs['LayerName']), order = list(df_sort_mean.index))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09f21d58",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "832033ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3121d45f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
